src/stream/src/common/table/state_table.rs

```diff
@@ -938,6 +937,34 @@ where
             .map(get_second))
     }
 
+    async fn iter_all_pk_and_val_with_pk_range(
+        &self,
+        pk_range: &(Bound<impl Row>, Bound<impl Row>),
+        prefetch_options: PrefetchOptions,
+    ) -> StreamExecutorResult<RowStreamWithPk<'_, S, SD>> {
```

Comment:
I'm still confused about why this compiles. 🥵 Doesn't it have a different type from `iter_key_and_val_with_pk_range`?

Reply:
Hmm what do you mean? Both have same type `StreamExecutorResult<RowStreamWithPk<'_, S, SD>>`.

The body of the function also merges all the streams inside into 1, so it seems expected? Did you mean the `pin` should change its type somehow? If so I'm not sure either 🤣

Reply:
> also merges all the streams inside into 1

`futures_async_stream` is essientially a [generator](https://doc.rust-lang.org/beta/unstable-book/language-features/generators.html), which is similar to an `async fn` or closure, which has different anonymous types for different instances. So merging these streams with `merge_sort` should generate a new type instead. 🤔

Reply:
i.e., `merge_sort(vec_it)` should have different type with `vec_it[0]`. Quite weird.

Reply:
I tried this
```
mod tests {
    type Op = impl Stream<Item = StorageResult<((), OwnedRow)>>;

    async fn iter() -> Op {
        futures::stream::iter(vec![Ok(((), OwnedRow::empty()))])
    }

    async fn wtf() -> Op {
        merge_sort(vec![Box::pin(iter().await)])
    }

    #[tokio::test]
    async fn wtff() {
        let mut it = iter().await;
        it = wtf().await;
    }
}
```
And triggered SIGSEGV.

Reply:
It looks dangerous 🥵 . Another reason tells us don't introduce this new interface for the state table.

Reply:
Add some reference here. https://rust-lang.github.io/rfcs/2515-type_alias_impl_trait.html

Reply:
> create a small reproduce and report an issue to rust

https://github.com/rust-lang/rust/issues/113314